lang/py3/avro/io.py

```diff
@@ -135,9 +135,17 @@ def Validate(expected_schema, datum):
     return any(Validate(union_branch, datum)
                for union_branch in expected_schema.schemas)
   elif schema_type in ['record', 'error', 'request']:
-    return (isinstance(datum, dict)
-        and all(Validate(field.type, datum.get(field.name))
-                for field in expected_schema.fields))
+    if not isinstance(datum, dict):
+        return False
+    expected_schema_field_names = set()
+    for field in expected_schema.fields:
+        expected_schema_field_names.add(field.name)
+        if not Validate(field.type, datum.get(field.name)):
+            return False
+    for datum_field in datum.keys():
```

Comment:
This block ensures all fields in the `datum` exist in the `expected_schema`. If they do not then the datum does not validate.

A stricter (but incompatible) change would be to create a `Null` type and require that all datum set all null fields to `Null` rather than relying on python's `None` type. The approach here is probably better for compatibility.